### PR TITLE
Pipe Resource Type

### PR DESCRIPTION
**Pull request by maxfischer2781**
Adds a Pipe resource type. This pull request includes:

* [x] new resource ``usim.Pipe`` to simulate pipes/fluxes of limited throughput,
* [x] unittests for ``usim.Pipe``,
* [x] basic documentation.

### CODE REVIEW

File: usim/_basics/pipe.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        .. code:: python3
+
+            await network.transfer(total=50e9)
+
+        :param total: absolute volume to transfer before resuming
+        :param throughput: maximum throughput
+        :return:
+        """
+        transferred = 0
```
Comment (eileen-kuehn) [inline usim/_basics/pipe.py]: I think it does make sense to ensure that the transfer is non-negative. What do you think?

```suggestion
        assert total > 0
        transferred = 0
```

Comment (maxfischer2781) [reply usim/_basics/pipe.py]: I've added asserts for all inputs now.

**Comment by maxfischer2781** [comment]
I have removed the ``Pipe.stream`` method for now. It would make more sense to stream in the background, but that has [some usability pitfalls](https://www.python.org/dev/peps/pep-0533/) that need working around.

I would prefer to preliminarily merge the current capabilities, since they are likely by far the most useful.

File: usim/_basics/pipe.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+from .._primitives.notification import Notification, suspend
+from .._core.loop import __LOOP_STATE__
+
+
+class Pipe:
+    """
+    Shared transport for resources with a limited total throughput
+
+    :parm throughput: limit of total throughput
```
Comment (eileen-kuehn) [inline usim/_basics/pipe.py]:
```suggestion
    :param throughput: limit of total throughput
```

File: usim_pytest/test_types/test_pipe.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+async def aenumerate(aiterable, start=0):
+    count = start
+    async for item in aiterable:
+        yield count, item
+        count += 1
+
+
+class TestPipe:
+    @via_usim
+    async def test_transfer_uncontested(self):
```
Comment (eileen-kuehn) [inline usim_pytest/test_types/test_pipe.py]: Your grammar correction is really picky ;)

```suggestion
    async def test_transfer_uncongested(self):
```

File: usim_pytest/test_types/test_pipe.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        assert (time == 4)
+        async with Scope() as scope:
+            scope.do(pipe.transfer(total=0.5, throughput=0.5))
+            scope.do(pipe.transfer(total=0.5, throughput=0.5))
+            scope.do(pipe.transfer(total=0.25, throughput=0.5))
+            scope.do(pipe.transfer(total=0.25, throughput=0.5))
+        assert (time == 5)
+
+    @via_usim
+    async def test_transfer_contested(self):
```
Comment (eileen-kuehn) [inline usim_pytest/test_types/test_pipe.py]:
```suggestion
    async def test_transfer_congested(self):
```

File: usim_pytest/test_types/test_pipe.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+async def aenumerate(aiterable, start=0):
+    count = start
+    async for item in aiterable:
+        yield count, item
+        count += 1
+
+
+class TestPipe:
```
Comment (eileen-kuehn) [inline usim_pytest/test_types/test_pipe.py]: Could you please also add a unit test for testing your assertions? So entering ``0`` and negative values?
